File: apps/api/src/services/ai/providers/OpenAIProvider.ts
/**
 * OpenAI Provider
 *
 * Real OpenAI SDK integration supporting:
 * - Chat completions (GPT-4, GPT-3.5-turbo)
 * - Embeddings (text-embedding-3-small, text-embedding-3-large)
 * - Whisper transcription
 * - Streaming support
 * - Vision models
 */

import OpenAI from 'openai';
import * as winston from 'winston';
import {
  IAIProvider,
  AIProviderType,
  ProviderCapabilities,
  ProviderConfig,
  ChatCompletionOptions,
  ChatCompletionResponse,
  EmbeddingOptions,
  EmbeddingResponse,
  TranscriptionOptions,
  TranscriptionResponse,
} from '../AIProviderFactory';

const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || 'info',
  defaultMeta: { service: 'openai-provider' },
  transports: [new winston.transports.Console()],
});

/**
 * OpenAI Provider Implementation
 */
export class OpenAIProvider implements IAIProvider {
  readonly type: AIProviderType = 'openai';
  readonly capabilities: ProviderCapabilities = {
    chat: true,
    embeddings: true,
    transcription: true,
    vision: true,
    streaming: true,
    functions: true,
  };

  private client: OpenAI;
  private config: Required<ProviderConfig>;

  constructor(config?: ProviderConfig) {
    // Load configuration with defaults
    this.config = {
      apiKey: config?.apiKey || process.env.OPENAI_API_KEY || '',
      baseURL: config?.baseURL || process.env.OPENAI_BASE_URL || 'https://api.openai.com/v1',
      model: config?.model || process.env.OPENAI_MODEL || 'gpt-4-turbo-preview',
      embeddingModel: config?.embeddingModel || process.env.OPENAI_EMBEDDING_MODEL || 'text-embedding-3-small',
      timeout: config?.timeout || 60000,
      maxRetries: config?.maxRetries || 3,
    };

    // Validate API key
    if (!this.config.apiKey) {
      throw new Error('OpenAI API key is required. Set OPENAI_API_KEY environment variable.');
    }

    // Initialize OpenAI client
    this.client = new OpenAI({
      apiKey: this.config.apiKey,
      baseURL: this.config.baseURL,
      timeout: this.config.timeout,
      maxRetries: this.config.maxRetries,
    });

    logger.info('OpenAI provider initialized', {
      baseURL: this.config.baseURL,
      model: this.config.model,
      embeddingModel: this.config.embeddingModel,
    });
  }

  /**
   * Get available OpenAI models
   */
  async getAvailableModels(): Promise<string[]> {
    try {
      const response = await this.client.models.list();
      const models = response.data
        .filter(m => m.id.startsWith('gpt-') || m.id.includes('embedding'))
        .map(m => m.id)
        .sort();

      logger.debug(`Retrieved ${models.length} OpenAI models`);
      return models;
    } catch (error) {
      logger.error('Failed to retrieve OpenAI models', {
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      // Return common models as fallback
      return [
        'gpt-4-turbo-preview',
        'gpt-4',
        'gpt-3.5-turbo',
        'text-embedding-3-small',
        'text-embedding-3-large',
      ];
    }
  }

  /**
   * Chat completion with GPT models
   */
  async chatCompletion(options: ChatCompletionOptions): Promise<ChatCompletionResponse> {
    const {
      messages,
      temperature = 0.7,
      maxTokens = 1000,
      stream = false,
      topP,
      frequencyPenalty,
      presencePenalty,
      stop,
    } = options;

    try {
      const startTime = Date.now();

      const response = await this.client.chat.completions.create({
        model: this.config.model,
        messages: messages as OpenAI.Chat.ChatCompletionMessageParam[],
        temperature,
        max_tokens: maxTokens,
        stream: false,
        top_p: topP,
        frequency_penalty: frequencyPenalty,
        presence_penalty: presencePenalty,
        stop,
      });

      const duration = Date.now() - startTime;

      const choice = response.choices[0];
      if (!choice?.message?.content) {
        throw new Error('No content in OpenAI response');
      }

      const result: ChatCompletionResponse = {
        content: choice.message.content,
        model: response.model,
        usage: response.usage
          ? {
              promptTokens: response.usage.prompt_tokens,
              completionTokens: response.usage.completion_tokens,
              totalTokens: response.usage.total_tokens,
            }
          : undefined,
        finishReason: choice.finish_reason || undefined,
      };

      logger.info('OpenAI chat completion successful', {
        model: response.model,
        duration,
        tokens: result.usage?.totalTokens,
        finishReason: result.finishReason,
      });

      return result;
    } catch (error) {
      logger.error('OpenAI chat completion failed', {
        model: this.config.model,
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }

  /**
   * Generate embeddings
   */
  async generateEmbeddings(options: EmbeddingOptions): Promise<EmbeddingResponse> {
    const { input, model } = options;
    const inputs = Array.isArray(input) ? input : [input];

    try {
      const startTime = Date.now();

      const response = await this.client.embeddings.create({
        model: model || this.config.embeddingModel,
        input: inputs,
      });

      const duration = Date.now() - startTime;

      const embeddings = response.data.map(item => item.embedding);

      const result: EmbeddingResponse = {
        embeddings,
        model: response.model,
        usage: response.usage
          ? {
              promptTokens: response.usage.prompt_tokens,
              totalTokens: response.usage.total_tokens,
            }
          : undefined,
      };

      logger.info('OpenAI embeddings generated', {
        model: response.model,
        count: embeddings.length,
        dimensions: embeddings[0]?.length || 0,
        duration,
      });

      return result;
    } catch (error) {
      logger.error('OpenAI embeddings generation failed', {
        model: model || this.config.embeddingModel,
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }

  /**
   * Transcribe audio using Whisper
   */
  async transcribeAudio(options: TranscriptionOptions): Promise<TranscriptionResponse> {
    const { audioBuffer, fileName, language, prompt } = options;

    try {
      const startTime = Date.now();

      // Create a File object from buffer
      const audioFile = new File([audioBuffer as any], fileName, {
        type: this.getAudioMimeType(fileName),
      }) as any;

      const response = await this.client.audio.transcriptions.create({
        file: audioFile,
        model: 'whisper-1',
        language,
        prompt,
        response_format: 'verbose_json',
      });

      const duration = Date.now() - startTime;

      const result: TranscriptionResponse = {
        text: response.text,
        language: (response as any).language,
        duration: (response as any).duration,
      };

      logger.info('OpenAI transcription successful', {
        fileName,
        textLength: result.text.length,
        language: result.language,
        duration,
      });

      return result;
    } catch (error) {
      logger.error('OpenAI transcription failed', {
        fileName,
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }

  /**
   * Test connection to OpenAI
   */
  async testConnection(): Promise<boolean> {
    try {
      await this.client.models.list();
      logger.debug('OpenAI connection test successful');
      return true;
    } catch (error) {
      logger.warn('OpenAI connection test failed', {
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      return false;
    }
  }

  /**
   * Get current configuration
   */
  getConfig(): ProviderConfig {
    return {
      apiKey: '***redacted***',
      baseURL: this.config.baseURL,
      model: this.config.model,
      embeddingModel: this.config.embeddingModel,
      timeout: this.config.timeout,
      maxRetries: this.config.maxRetries,
    };
  }

  /**
   * Helper to determine audio MIME type from filename
   */
  private getAudioMimeType(fileName: string): string {
    const ext = fileName.toLowerCase().split('.').pop();
    const mimeTypes: Record<string, string> = {
      mp3: 'audio/mpeg',
      mp4: 'audio/mp4',
      m4a: 'audio/mp4',
      wav: 'audio/wav',
      webm: 'audio/webm',
      ogg: 'audio/ogg',
    };
    return mimeTypes[ext || ''] || 'audio/mpeg';
  }

  /**
   * Chat completion with streaming (for future use)
   */
  async *chatCompletionStream(
    options: ChatCompletionOptions
  ): AsyncGenerator<string, void, unknown> {
    const {
      messages,
      temperature = 0.7,
      maxTokens = 1000,
      topP,
      frequencyPenalty,
      presencePenalty,
      stop,
    } = options;

    try {
      const stream = await this.client.chat.completions.create({
        model: this.config.model,
        messages: messages as OpenAI.Chat.ChatCompletionMessageParam[],
        temperature,
        max_tokens: maxTokens,
        stream: true,
        top_p: topP,
        frequency_penalty: frequencyPenalty,
        presence_penalty: presencePenalty,
        stop,
      });

      for await (const chunk of stream) {
        const content = chunk.choices[0]?.delta?.content;
        if (content) {
          yield content;
        }
      }
    } catch (error) {
      logger.error('OpenAI streaming completion failed', {
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }
}

File: apps/ai-service/app/services/providers/openai_provider.py
"""
OpenAI Provider Implementation
Supports: GPT-4, GPT-3.5, Whisper, DALL-E, Embeddings
"""

import asyncio
import logging
from typing import List, Dict, Any, Optional, AsyncIterator
import openai
from openai import OpenAI, AsyncOpenAI
import os
import time

from .base_provider import (
    AIProvider,
    ProviderType,
    ProviderConfig,
    ModelCapability,
    ModelInfo,
    TranscriptionRequest,
    TranscriptionResponse,
    ChatRequest,
    ChatResponse,
    ChatMessage,
    EmbeddingRequest,
    EmbeddingResponse,
    VisionRequest,
    VisionResponse,
)

logger = logging.getLogger(__name__)


class OpenAIProvider(AIProvider):
    """OpenAI API provider implementation"""

    def __init__(self, config: ProviderConfig):
        super().__init__(config)

        # Initialize OpenAI clients
        api_key = config.api_key or os.getenv("OPENAI_API_KEY")
        if not api_key:
            raise ValueError("OpenAI API key not provided")

        self.client = OpenAI(
            api_key=api_key,
            organization=config.organization,
            base_url=config.api_base,
            timeout=config.timeout,
            max_retries=config.max_retries,
        )

        self.async_client = AsyncOpenAI(
            api_key=api_key,
            organization=config.organization,
            base_url=config.api_base,
            timeout=config.timeout,
            max_retries=config.max_retries,
        )

        # Model registry
        self.models = {
            "gpt-4-turbo": ModelInfo(
                model_id="gpt-4-turbo-preview",
                name="GPT-4 Turbo",
                provider=ProviderType.OPENAI,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION],
                context_window=128000,
                max_output_tokens=4096,
                cost_per_1k_input=0.01,
                cost_per_1k_output=0.03,
                supports_streaming=True,
                supports_function_calling=True,
            ),
            "gpt-4": ModelInfo(
                model_id="gpt-4",
                name="GPT-4",
                provider=ProviderType.OPENAI,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION],
                context_window=8192,
                max_output_tokens=4096,
                cost_per_1k_input=0.03,
                cost_per_1k_output=0.06,
                supports_streaming=True,
                supports_function_calling=True,
            ),
            "gpt-3.5-turbo": ModelInfo(
                model_id="gpt-3.5-turbo",
                name="GPT-3.5 Turbo",
                provider=ProviderType.OPENAI,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION],
                context_window=16385,
                max_output_tokens=4096,
                cost_per_1k_input=0.0005,
                cost_per_1k_output=0.0015,
                supports_streaming=True,
                supports_function_calling=True,
            ),
            "whisper-1": ModelInfo(
                model_id="whisper-1",
                name="Whisper",
                provider=ProviderType.OPENAI,
                capabilities=[ModelCapability.TRANSCRIPTION],
                context_window=0,
                max_output_tokens=0,
                cost_per_1k_input=0.006,  # per minute
                cost_per_1k_output=0.0,
                supports_streaming=False,
            ),
            "text-embedding-3-large": ModelInfo(
                model_id="text-embedding-3-large",
                name="Text Embedding 3 Large",
                provider=ProviderType.OPENAI,
                capabilities=[ModelCapability.EMBEDDING],
                context_window=8191,
                max_output_tokens=0,
                cost_per_1k_input=0.00013,
                cost_per_1k_output=0.0,
                supports_streaming=False,
            ),
        }

        logger.info(f"✅ OpenAI Provider initialized ({len(self.models)} models)")

    async def transcribe(self, request: TranscriptionRequest) -> TranscriptionResponse:
        """Transcribe audio using OpenAI Whisper"""
        try:
            start_time = time.time()

            with open(request.audio_path, "rb") as audio_file:
                response = await self.async_client.audio.transcriptions.create(
                    file=audio_file,
                    model="whisper-1",
                    language=request.language,
                    response_format="verbose_json",
                    timestamp_granularities=["segment"] if request.enable_timestamps else None,
                )

            processing_time = time.time() - start_time

            # Convert segments
            segments = []
            if hasattr(response, "segments") and response.segments:
                segments = [
                    {
                        "id": i,
                        "start": seg.get("start", 0.0),
                        "end": seg.get("end", 0.0),
                        "text": seg.get("text", ""),
                        "confidence": 0.9,  # OpenAI doesn't provide confidence
                    }
                    for i, seg in enumerate(response.segments)
                ]

            return TranscriptionResponse(
                text=response.text,
                segments=segments,
                language=response.language if hasattr(response, "language") else request.language or "en",
                duration=response.duration if hasattr(response, "duration") else 0.0,
                confidence=0.9,
                provider="openai",
                model="whisper-1",
                processing_time=processing_time,
            )

        except Exception as e:
            logger.error(f"OpenAI transcription failed: {e}")
            raise

    async def chat_completion(self, request: ChatRequest) -> ChatResponse:
        """Generate chat completion"""
        try:
            model = request.model or "gpt-4-turbo-preview"

            # Convert messages
            messages = [
                {"role": msg.role, "content": msg.content}
                for msg in request.messages
            ]

            response = await self.async_client.chat.completions.create(
                model=model,
                messages=messages,
                temperature=request.temperature,
                max_tokens=request.max_tokens,
                top_p=request.top_p,
                frequency_penalty=request.frequency_penalty,
                presence_penalty=request.presence_penalty,
                stop=request.stop,
                stream=False,
                functions=request.functions,
            )

            choice = response.choices[0]

            return ChatResponse(
                content=choice.message.content or "",
                model=response.model,
                provider="openai",
                finish_reason=choice.finish_reason,
                usage={
                    "prompt_tokens": response.usage.prompt_tokens,
                    "completion_tokens": response.usage.completion_tokens,
                    "total_tokens": response.usage.total_tokens,
                },
                function_call=choice.message.function_call.__dict__ if choice.message.function_call else None,
            )

        except Exception as e:
            logger.error(f"OpenAI chat completion failed: {e}")
            raise

    async def chat_completion_stream(self, request: ChatRequest) -> AsyncIterator[str]:
        """Generate streaming chat completion"""
        try:
            model = request.model or "gpt-4-turbo-preview"

            messages = [
                {"role": msg.role, "content": msg.content}
                for msg in request.messages
            ]

            stream = await self.async_client.chat.completions.create(
                model=model,
                messages=messages,
                temperature=request.temperature,
                max_tokens=request.max_tokens,
                top_p=request.top_p,
                stream=True,
            )

            async for chunk in stream:
                if chunk.choices[0].delta.content:
                    yield chunk.choices[0].delta.content

        except Exception as e:
            logger.error(f"OpenAI streaming failed: {e}")
            raise

    async def generate_embedding(self, request: EmbeddingRequest) -> EmbeddingResponse:
        """Generate embeddings"""
        try:
            model = request.model or "text-embedding-3-large"

            # Handle single string or list
            input_texts = [request.input] if isinstance(request.input, str) else request.input

            response = await self.async_client.embeddings.create(
                model=model,
                input=input_texts,
            )

            embeddings = [item.embedding for item in response.data]

            return EmbeddingResponse(
                embeddings=embeddings,
                model=response.model,
                provider="openai",
                usage={
                    "prompt_tokens": response.usage.prompt_tokens,
                    "total_tokens": response.usage.total_tokens,
                },
            )

        except Exception as e:
            logger.error(f"OpenAI embedding failed: {e}")
            raise

    async def vision_completion(self, request: VisionRequest) -> VisionResponse:
        """Process vision request with GPT-4 Vision"""
        try:
            # TODO: Implement GPT-4 Vision support
            raise NotImplementedError("GPT-4 Vision not yet implemented")

        except Exception as e:
            logger.error(f"OpenAI vision failed: {e}")
            raise

    def list_models(self) -> List[ModelInfo]:
        """List available models"""
        return list(self.models.values())

    async def health_check(self) -> bool:
        """Check OpenAI API health"""
        try:
            # Simple test request
            await self.async_client.models.list()
            return True
        except Exception as e:
            logger.warning(f"OpenAI health check failed: {e}")
            return False

    def get_capabilities(self) -> List[ModelCapability]:
        """Get provider capabilities"""
        return [
            ModelCapability.CHAT,
            ModelCapability.TEXT_GENERATION,
            ModelCapability.TRANSCRIPTION,
            ModelCapability.EMBEDDING,
            ModelCapability.VISION,
        ]

    async def estimate_cost(
        self,
        request: Union[ChatRequest, TranscriptionRequest],
        model: Optional[str] = None
    ) -> Dict[str, float]:
        """Estimate cost for request"""
        if isinstance(request, ChatRequest):
            model_id = model or request.model or "gpt-4-turbo-preview"
            model_info = self.models.get(model_id.split("/")[-1])

            if not model_info:
                return {"input_cost": 0.0, "output_cost": 0.0, "total_cost": 0.0}

            # Rough token estimation (4 chars ≈ 1 token)
            total_chars = sum(len(msg.content) for msg in request.messages)
            estimated_input_tokens = total_chars // 4
            estimated_output_tokens = request.max_tokens or 500

            input_cost = (estimated_input_tokens / 1000) * model_info.cost_per_1k_input
            output_cost = (estimated_output_tokens / 1000) * model_info.cost_per_1k_output
            total_cost = input_cost + output_cost

            return {
                "input_cost": round(input_cost, 6),
                "output_cost": round(output_cost, 6),
                "total_cost": round(total_cost, 6),
                "estimated_input_tokens": estimated_input_tokens,
                "estimated_output_tokens": estimated_output_tokens,
            }

        elif isinstance(request, TranscriptionRequest):
            # Whisper costs $0.006 per minute
            # Estimate duration from file size (rough: 1MB ≈ 1 minute)
            import os
            file_size_mb = os.path.getsize(request.audio_path) / (1024 * 1024)
            estimated_minutes = max(1, file_size_mb)
            total_cost = estimated_minutes * 0.006

            return {
                "input_cost": round(total_cost, 6),
                "output_cost": 0.0,
                "total_cost": round(total_cost, 6),
                "estimated_minutes": estimated_minutes,
            }

        return {"input_cost": 0.0, "output_cost": 0.0, "total_cost": 0.0}
